### PR TITLE
Added methods for Sequence and Struct

### PR DESCRIPTION
**Pull request by desaikd**
*Description of changes:*
This PR works on creating methods for `Sequence` and `Struct` in Value API.

*Completed Tasks:*
- Added `Eq`, `PartialEq` and `Hash` implementations for` Borrowed/OwnedSymbolTokens`.
- Added Sequence methods: `get`, `len`, `is_empty`, `pop`, `push`
- Added Struct methods: `get`, `get_all`
- Struct uses a `HashMap` with `(key, values) -> (SymbolToken, Vec<Element>)`

*TODO:*
- Remove usage of `unwrap()` as it will panic on `None`.
- Add tests for this implementation

*Base PR:* https://github.com/amzn/ion-rust/pull/131

### CODE REVIEW

File: src/value/owned.rs [hunk 1]
```diff
@@ -120,12 +168,25 @@ impl Struct for OwnedStruct {
...
+    }
+
+    // TODO remove usage of unwrap()[unwrap] as it will panic for None
+    //  [unwrap] https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap
+    fn get_all<'a>(
+        &'a self,
+        field_name: &'a Self::FieldName,
+    ) -> Box<dyn Iterator<Item = &'a Self::Element> + 'a> {
+        Box::new(self.fields.get(field_name).unwrap().iter())
+    }
```
Comment (almann) [inline src/value/owned.rs:189]: One thing to consider is that `Option<T>` is a container so you can iterate/map over it.  Consider this [snippet][1]:

```rust
fn mk_iter<'a, T>(opt: Option<&'a Vec<T>>) -> Box<dyn Iterator<Item = &'a T> + 'a> {
    Box::new(opt.into_iter().flat_map(|v| v.iter()))
}

fn to_string_vec<'a, I: Iterator<Item = &'a String> + 'a>(iter: I) -> Vec<String> {
    iter.map(|e| e.to_owned()).collect()
}

fn main() {
    let vals = vec!["a", "b", "c"].iter().map(|s| s.to_string()).collect();
    let some_vals = Some(&vals);
    let new_some_vals: Vec<String> = to_string_vec(mk_iter(some_vals));
    assert_eq!(vals, new_some_vals);

    let none_vals: Option<&Vec<String>> = None;
    let new_none_vals: Vec<String> = to_string_vec(mk_iter(none_vals));
    assert!(new_none_vals.is_empty());
}
```

[1]: https://play.rust-lang.org/?version=stable&edition=2018&gist=f64e8eadf682f82480a86da03549c4f8

File: src/value/borrowed.rs [hunk 2]
```diff
@@ -79,6 +81,27 @@ impl<'val> SymbolToken for BorrowedSymbolToken<'val> {
...
+            true
+        } else if self.source == other.source {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl<'val> Eq for BorrowedSymbolToken<'val> {}
```
Comment (almann) [inline src/value/borrowed.rs:103]: See my comments in the owned variants for this.

File: src/value/borrowed.rs [hunk 3]
```diff
@@ -119,12 +167,25 @@ impl<'val> Struct for BorrowedStruct<'val> {
...
+
+    // TODO remove usage of unwrap()[unwrap] as it will panic for None
+    //  [unwrap] https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap
+    fn get_all<'a>(
+        &'a self,
+        field_name: &'a Self::FieldName,
+    ) -> Box<dyn Iterator<Item = &'a Self::Element> + 'a> {
+        Box::new(self.fields.get(field_name).unwrap().iter())
+    }
 }
```
Comment (almann) [inline src/value/borrowed.rs:189]: Likewise--see my comment in the owned variants for this.

**Review by almann — CHANGES_REQUESTED** [review]
Good start, but I think there are some things to really consider in the implementation around hashing/equality and `struct` layout vis-a-vis `HashMap`.

Also please make sure to use `///` for `rustdoc` friendly comments.

File: src/value/mod.rs [hunk 4]
```diff
@@ -318,7 +318,20 @@ pub trait Sequence {
...
+    fn len(&self) -> usize;
+
+    // check if the sequence is empty or not
+    fn is_empty(&self) -> bool;
+
+    // returns the last element from the sequence
+    fn pop(&mut self) -> Self::Element;
+
+    // Appends the given value e to the end of the Array
+    fn push(&mut self, e: Self::Element);
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/mod.rs:334]: Yes. I think we can work on the mutable operations later once we figure out a generic way to do it. I will place these methods into TODO for now.

File: src/value/mod.rs [hunk 5]
```diff
@@ -335,7 +348,14 @@ pub trait Struct {
     /// [gat]: https://rust-lang.github.io/rfcs/1598-generic_associated_types.html
     fn iter<'a>(
         &'a self,
-    ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, &'a Self::Element)> + 'a>;
+    ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, Option<&'a Self::Element>)> + 'a>;
+
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/mod.rs:352]: Sorry. This seems incorrect I guess. I would assume that `fields`  would be a `HashMap` with `(field_name, values)` where `values` can not be empty(returning `None` here) as a field would only be added if it has at least one value. does that answer your question?

File: src/value/mod.rs [hunk 6]
```diff
@@ -335,7 +348,14 @@ pub trait Struct {
...
+    // gets the first value for the given field name in struct
+    fn get(&self, field_name: &Self::FieldName) -> Option<&Self::Element>;
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/mod.rs:354]: I will try to add a generic implementation for it which would allow `&str` `&String` lookup.

File: src/value/owned.rs [hunk 7]
```diff
@@ -80,6 +82,27 @@ impl SymbolToken for OwnedSymbolToken {
     }
 }
 
+impl Hash for OwnedSymbolToken {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.text.hash(state);
+        self.source.hash(state);
+    }
+}
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/owned.rs:90]: I could have added this into TODO. It needs to be changed according to data model.

File: src/value/owned.rs [hunk 8]
```diff
@@ -80,6 +82,27 @@ impl SymbolToken for OwnedSymbolToken {
...
+
+impl PartialEq for OwnedSymbolToken {
+    fn eq(&self, other: &Self) -> bool {
+        if self.text == other.text {
+            true
+        } else if self.source == other.source {
+            true
+        } else {
+            false
+        }
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/owned.rs:100]: Yes. My intent here is that  `hash` and `eq` should follow the data model.

File: src/value/owned.rs (hunk 1 above)
Comment (desaikd) [reply src/value/owned.rs:189]: Thanks! This works fine for `get_all`.

File: src/value/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (almann) [reply src/value/mod.rs:352]: I might be not asking the question well. Let me rephrase, There seems to be a change here to the iterator in which the items are pairs of `&Self::FieldName` and `Option<&Element>`.  Why `Option<&Self::Element>` or more specifically what does it mean to iterate over a struct and get some field name with `None` instead of `Some<&Self::Element>`?

**Comment by almann** [comment]
I've been thinking a lot about the issues of using something like `HashMap<SymbolToken, Vec<Element>>`  as the `Struct` representation and I think it leads to a lot of headache around the weird cases of `SymbolToken`.

I think a cleaner approach might be to have an index of `HashMap<String, Vec<(SymbolToken, Element)>>` and a `Vec<(SymbolToken, Element)>` to capture any symbol tokens that happen to have no text representation.  You could then use `flat_map` and `chain` to combine these two for the iterator over the field members.

It is unfortunate that we need a copy of the string and the token, but at least in the borrowed version the cost should just extra reference of `&str`.

**Comment by almann** [comment]
For reference, this makes progress on #114.

File: src/value/owned.rs [hunk 9]
```diff
@@ -168,24 +154,42 @@ impl Struct for OwnedStruct {
...
+        // convert &(k, v) -> (&k, &v)
+        // flattens the fields_with_text_key HashMap and chains with fields_with_no_text_key
+        // to return all fields with iterator
+        let all_fields_vec: Vec<&(OwnedSymbolToken, OwnedElement)> = self
+            .fields_with_text_key
+            .values()
+            .flat_map(|v| v)
+            .into_iter()
+            .chain(self.fields_with_no_text_key.iter())
+            .collect();
```
Comment (almann) [inline src/value/owned.rs:167]: Similar question--is the `collect` necessary?

**Review by almann — COMMENTED** [review]
Looks close--though I do have a couple questions below and some style questions.

File: src/value/borrowed.rs [hunk 10]
```diff
@@ -137,27 +115,36 @@ impl<'val> Sequence for BorrowedSequence<'val> {
...
+    fields_with_no_text_key: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>,
 }
 
 impl<'val> BorrowedStruct<'val> {
-    pub fn new(fields: HashMap<BorrowedSymbolToken<'val>, Vec<BorrowedElement<'val>>>) -> Self {
-        Self { fields }
+    // TODO remove constructor and instead add from_iter function to HashMap and Vec from iterator
+    fn new(
+        fields_with_text: HashMap<
+            std::string::String,
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/borrowed.rs:139]: No I will modify it with all the other changes here.

File: src/value/borrowed.rs [hunk 11]
```diff
@@ -137,27 +115,36 @@ impl<'val> Sequence for BorrowedSequence<'val> {
...
 }
 
 /// A borrowed implementation of [`Struct`]
 #[derive(Debug, Clone)]
 pub struct BorrowedStruct<'val> {
-    // TODO model actual map indexing for the struct (maybe as a variant type)
-    //      otherwise struct field lookup will be O(N)
-    fields: HashMap<BorrowedSymbolToken<'val>, Vec<BorrowedElement<'val>>>,
+    fields_with_text_key: HashMap<String, Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>>,
+    fields_with_no_text_key: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>,
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/value/borrowed.rs:132]: yes I think those fits better as they are less verbose.

File: src/value/borrowed.rs [hunk 12]
```diff
@@ -158,20 +146,20 @@ impl<'val> Struct for BorrowedStruct<'val> {
...
-        }))
+        Box::new(
+            self.text_fields
+                .values()
+                .flat_map(|v| v)
+                .into_iter()
+                .chain(self.no_text_fields.iter())
+                .map(|kv| match &kv {
+                    (k, v) => (k, v),
+                }),
```
Comment (almann) [inline src/value/borrowed.rs:157]: *Minor*: You should be able to do the de-structuring in the closure argument FWIW

```rust
map(|(k, v)| (k, v))
```

Just eliminates the extra `match` expression.

File: src/value/owned.rs [hunk 13]
```diff
@@ -158,20 +149,20 @@ impl Struct for OwnedStruct {
...
-        }))
+        Box::new(
+            self.text_fields
+                .values()
+                .flat_map(|v| v)
+                .into_iter()
+                .chain(self.no_text_fields.iter())
+                .map(|kv| match &kv {
+                    (k, v) => (k, v),
+                }),
```
Comment (almann) [inline src/value/owned.rs:160]: *Minor*: See above.

**Review by almann — APPROVED** [review]
Looks good--one basically minor comment about de-structuring in the closure parameter to avoid an extra `match`, but really this is ready to :ship:

**Review by almann — APPROVED** [review]
:rocket:

File: src/value/borrowed.rs [hunk 14]
```diff
@@ -97,19 +98,41 @@ impl<'val> Sequence for BorrowedSequence<'val> {
...
         Box::new(self.children.iter())
     }
+
+    fn get(&self, index: usize) -> Option<&Self::Element> {
+        if index > self.children.len() {
+            None
+        } else {
+            Some(&self.children[index])
+        }
+    }
```
Comment (zslayton) [inline src/value/borrowed.rs:108]: I'm late to the party here, but this method body can be replaced with `self.children.get(index)`, which already returns an `Option` wrapping the value at that index.
